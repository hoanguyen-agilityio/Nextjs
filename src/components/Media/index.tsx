'use client';

// Third party
import clsx from 'clsx';
import { useRef, useState, useEffect } from 'react';

// Components
import { ButtonCustom, ImageCustom, InputCustom } from '@/components';
import { ImageIcon } from '@/icons';
import { MESSAGE } from '@/constants';

interface ImageUploaderProps {
  mode: 'add' | 'edit' | 'detail';
  onImagesChange?: (images: string[]) => void;
  data?: { img: string[] };
}

const Media = ({ mode, onImagesChange, data }: ImageUploaderProps) => {
  const [images, setImages] = useState<string[]>(data?.img || []);
  const [url, setUrl] = useState<string>('');
  const [showModal, setShowModal] = useState<boolean>(false);
  const fileInputRef = useRef<HTMLInputElement | null>(null);
  const [errorMessage, setErrorMessage] = useState<string | null>(null);

  useEffect(() => {
    if (data?.img) {
      setImages(data.img);
    }
  }, [data?.img]);

  // Handle file upload (select or drop)
  const handleFileUpload = (files: FileList | null) => {
    if (files) {
      const newImages = Array.from(files).map((file) =>
        URL.createObjectURL(file),
      );
      const updatedImages = [...images, ...newImages];

      if (updatedImages.length > 10) {
        setErrorMessage(MESSAGE.IMAGE_OVERLOAD);
        return;
      }

      setImages(updatedImages);
      setErrorMessage(null);
      onImagesChange?.(updatedImages);
    }
  };

  const handleDrop = (event: React.DragEvent<HTMLDivElement>) => {
    event.preventDefault();
    handleFileUpload(event.dataTransfer.files);
  };

  const handleDragOver = (event: React.DragEvent<HTMLDivElement>) => {
    event.preventDefault();
  };

  const handleUrlSubmit = () => {
    if (url) {
      if (images.length > 10) {
        setErrorMessage(MESSAGE.IMAGE_OVERLOAD);
        return;
      }

      const updatedImages = [...images, url];
      setImages(updatedImages);
      setErrorMessage(null);
      onImagesChange?.(updatedImages);
      setUrl('');
      setShowModal(false);
    }
  };

  const handleImageDelete = (index: number) => {
    const updatedImages = images.filter((_, i) => i !== index);
    setImages(updatedImages);
    setErrorMessage(null);
    onImagesChange?.(updatedImages);
    URL.revokeObjectURL(images[index]);
  };

  const handleAreaClick = () => {
    setShowModal(true);
  };

  const handleFileInputClick = () => {
    if (fileInputRef.current) {
      fileInputRef.current.click();
    }
  };

  const handleFileInputChange = (files: FileList | null) => {
    handleFileUpload(files);
    setShowModal(false);
  };

  return (
    <div className="mt-3xl">
      <label
        className="font-semibold text-lg text-black-900 dark:text-white"
        htmlFor="Media"
      >
        Media
      </label>

      {mode !== 'detail' && (
        <div
          className="upload-area border-2 border-dashed border-grayBlue-400 rounded-md p-8 flex flex-col items-center justify-center text-ratio-200 dark:text-grayBlue-400 cursor-pointer gap-3 mt-3"
          onDrop={handleDrop}
          onDragOver={handleDragOver}
          onClick={handleAreaClick}
        >
          <ImageIcon width="60px" height="60px" />
          <p className="text-center font-normal text-xsm text-ratio-200 dark:text-grayBlue-400">
            Drop your images here, or click to browse 1600 x 1200 (4:3) <br />
            recommended, up to 10MB each.
          </p>
          <InputCustom
            type="file"
            accept="image/*"
            multiple
            onChange={(e) => handleFileUpload(e.target.files)}
            ref={fileInputRef}
            className="hidden"
          />
        </div>
      )}

      <div className="image-previews mt-4 grid grid-cols-2 gap-2">
        {images.map((src, index) => (
          <div key={index} className="relative group">
            <ImageCustom
              src={src}
              alt={`Preview ${index}`}
              classNames={{
                img: 'w-full h-32 object-cover rounded-md',
                wrapper: 'bg-cover',
              }}
            />
            {mode !== 'detail' && (
              <ButtonCustom
                type="button"
                onClick={() => handleImageDelete(index)}
                className="absolute top-2 right-2 bg-red-500 text-white rounded-full w-6 h-6 flex items-center justify-center opacity-0 group-hover:opacity-100 transition-opacity"
                aria-label="Delete image"
              >
                &times;
              </ButtonCustom>
            )}
          </div>
        ))}
      </div>

      <p
        className={clsx(
          'font-normal text-xsm text-ratio-200 dark:text-grayBlue-400 mt-3',
          { 'text-red-500': errorMessage },
        )}
      >
        {errorMessage ||
          'Add up to 10 images to your product. Used to represent your product during checkout, in email, social sharing, and more.'}
      </p>

      {showModal && (
        <div className="modal fixed inset-0 flex items-center justify-center bg-black bg-opacity-50 z-50">
          <div className="modal-content bg-white p-4 rounded-md shadow-md">
            <h4 className="text-lg font-medium mb-2">Add Image</h4>
            <div className="flex flex-col mb-2">
              <ButtonCustom
                type="button"
                color="primary"
                radius="sm"
                onClick={handleFileInputClick}
                className="rounded-md px-4 py-2 mb-2"
              >
                Choose Image from Computer
              </ButtonCustom>
              <InputCustom
                type="file"
                accept="image/*"
                multiple
                onChange={(e) => handleFileInputChange(e.target.files)}
                ref={fileInputRef}
                className="hidden"
              />
              <InputCustom
                type="text"
                placeholder="Enter image URL"
                value={url}
                onChange={(e) => setUrl(e.target.value)}
                className="border border-gray-300 rounded-md p-2 w-full"
              />
            </div>
            <div className="flex justify-end">
              <ButtonCustom
                onClick={handleUrlSubmit}
                color="primary"
                className={clsx('rounded-md px-4 py-2 mr-2', {
                  'cursor-not-allowed opacity-50': !url.trim(),
                })}
                disabled={!url.trim()}
              >
                Add
              </ButtonCustom>
              <ButtonCustom
                onClick={() => setShowModal(false)}
                className="bg-gray-300 text-black rounded-md px-4 py-2"
              >
                Cancel
              </ButtonCustom>
            </div>
          </div>
        </div>
      )}
    </div>
  );
};

export default Media;
