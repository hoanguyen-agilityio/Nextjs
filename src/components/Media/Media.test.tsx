import { render, fireEvent, screen } from '@testing-library/react';
import Media from '.';

describe('Media Component', () => {
  const mockOnImagesChange = jest.fn();

  const renderMedia = (
    mode: 'add' | 'edit' | 'detail',
    data?: { img: string[] },
  ) =>
    render(
      <Media mode={mode} onImagesChange={mockOnImagesChange} data={data} />,
    );

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders with default type', () => {
    const { container } = render(<Media mode="add" />);
    expect(container).toMatchSnapshot();
  });

  test('does not render the upload area in "detail" mode', () => {
    renderMedia('detail');
    expect(
      screen.queryByText(/Drop your images here/i),
    ).not.toBeInTheDocument();
  });

  test('displays modal when the upload area is clicked', () => {
    renderMedia('add');

    fireEvent.click(screen.getByText(/Drop your images here/i));

    expect(screen.getByText(/Add Image/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Enter image URL')).toBeInTheDocument();
  });

  test('closes modal on "Cancel" button click', () => {
    renderMedia('add');

    fireEvent.click(screen.getByText(/Drop your images here/i));
    const cancelButton = screen.getByText(/Cancel/i);
    fireEvent.click(cancelButton);

    expect(screen.queryByText(/Add Image/i)).not.toBeInTheDocument();
  });

  test('renders correctly in "add" mode with no initial data', () => {
    const { container } = renderMedia('add');
    expect(container).toMatchSnapshot();
    expect(screen.getByText(/Drop your images here/i)).toBeInTheDocument();
  });

  test('renders images passed in the "data" prop', () => {
    const data = { img: ['image1.jpg', 'image2.jpg'] };
    renderMedia('edit', data);

    const images = screen.getAllByRole('img');
    expect(images).toHaveLength(2);
    expect(images[0]).toHaveAttribute('src', 'image1.jpg');
    expect(images[1]).toHaveAttribute('src', 'image2.jpg');
  });
});
