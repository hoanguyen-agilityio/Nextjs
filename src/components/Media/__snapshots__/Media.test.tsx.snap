// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Media Component renders correctly in "add" mode with no initial data 1`] = `
<div>
  <div
    class="mt-3xl"
  >
    <label
      class="font-semibold text-lg text-black-900 dark:text-white"
      for="Media"
    >
      Media
    </label>
    <div
      class="upload-area border-2 border-dashed border-grayBlue-400 rounded-md p-8 flex flex-col items-center justify-center text-ratio-200 dark:text-grayBlue-400 cursor-pointer gap-3 mt-3"
    >
      <svg
        fill="none"
        height="60px"
        viewBox="0 0 61 60"
        width="60px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="30.0996"
          cy="30"
          fill="#99B2C6"
          r="30"
        />
        <path
          class="fill-white dark:fill-black-900"
          clip-rule="evenodd"
          d="M26.7216 40.5996C21.1996 40.5996 18.8406 38.2296 18.8406 32.6836V26.5566C18.8406 21.0106 21.2006 18.6416 26.7216 18.6416H32.8215C38.3435 18.6416 40.7025 21.0106 40.7025 26.5566V31.4306C40.7172 31.5167 40.7172 31.6046 40.7025 31.6906V32.6836C40.7025 38.2296 38.3435 40.5996 32.8215 40.5996H26.7216ZM26.7216 39.0676H32.8215C37.5095 39.0676 39.1776 37.3926 39.1776 32.6837V31.9056L37.7955 30.7136C37.2455 30.2743 36.4646 30.2743 35.9146 30.7136L31.6846 34.3596C30.5638 35.2887 28.9404 35.2887 27.8196 34.3596L27.4836 34.0596C26.976 33.6567 26.2675 33.6237 25.7245 33.9777L21.2656 36.9836C22.1995 38.4756 23.9235 39.0726 26.7216 39.0726V39.0676ZM20.3655 26.5566V32.6837C20.3453 33.6467 20.446 34.6086 20.6656 35.5466L24.8806 32.7047C25.9948 31.9654 27.4619 32.0442 28.4905 32.8986L28.8265 33.1986C29.3729 33.6378 30.1512 33.6378 30.6975 33.1986L34.9275 29.5526C36.048 28.6227 37.6719 28.6227 38.7925 29.5526L39.1775 29.8846V26.5596C39.1775 21.8516 37.5095 20.1766 32.8215 20.1766H26.7215C22.0336 20.1786 20.3666 21.8536 20.3666 26.5616L20.3655 26.5566ZM23.9255 25.5356C23.9305 23.9904 25.1863 22.7414 26.7314 22.7447C28.2767 22.748 29.5271 24.0024 29.5255 25.5477C29.5238 27.0929 28.2707 28.3447 26.7255 28.3447C25.982 28.3447 25.269 28.049 24.7438 27.5228C24.2186 26.9966 23.9242 26.2831 23.9255 25.5396V25.5356ZM25.4515 25.5356C25.4543 26.2369 26.0246 26.8035 26.7259 26.8016C27.4272 26.7998 27.9944 26.2303 27.9935 25.5289C27.9926 24.8276 27.4238 24.2596 26.7225 24.2596C26.3842 24.2607 26.0602 24.3962 25.8218 24.6363C25.5834 24.8763 25.4502 25.2013 25.4515 25.5397V25.5356Z"
          fill-rule="evenodd"
        />
      </svg>
      <p
        class="text-center font-normal text-xsm text-ratio-200 dark:text-grayBlue-400"
      >
        Drop your images here, or click to browse 1600 x 1200 (4:3) 
        <br />
        recommended, up to 10MB each.
      </p>
      <div
        class="group flex-col data-[hidden=true]:hidden w-full relative justify-end hidden"
        data-filled="true"
        data-filled-within="true"
        data-has-value="true"
        data-slot="base"
      >
        <div
          class="flex flex-col w-xl h-14"
          data-slot="main-wrapper"
        >
          <div
            class="relative tap-highlight-transparent flex-row shadow-sm gap-3 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 min-h-10 transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background bg-gray-50 flex items-center space-x-2 h-full placeholder:text-ratio-300 dark:bg-gray-700 dark: text-ratio-700 dark:placeholder:text-ratio-700 focus-within:ring-0 py-4 px-6 w-full rounded-lg"
            data-slot="input-wrapper"
            style="cursor: text;"
          >
            <div
              class="w-full items-center h-full box-border flex"
              data-slot="inner-wrapper"
            >
              <input
                accept="image/*"
                aria-label=" "
                class="w-full bg-transparent !outline-none focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text group-data-[has-value=true]:text-default-foreground placeholder:text-ratio-300 dark:placeholder:text-ratio-700 dark: text-ratio-700 flex-grow focus:outline-none focus:ring-0 text-base font-semibold"
                data-slot="input"
                id="react-aria-:r13:"
                multiple=""
                type="file"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="image-previews mt-4 grid grid-cols-2 gap-2"
    />
    <p
      class="font-normal text-xsm text-ratio-200 dark:text-grayBlue-400 mt-3"
    >
      Add up to 10 images to your product. Used to represent your product during checkout, in email, social sharing, and more.
    </p>
  </div>
</div>
`;

exports[`Media Component renders with default type 1`] = `
<div>
  <div
    class="mt-3xl"
  >
    <label
      class="font-semibold text-lg text-black-900 dark:text-white"
      for="Media"
    >
      Media
    </label>
    <div
      class="upload-area border-2 border-dashed border-grayBlue-400 rounded-md p-8 flex flex-col items-center justify-center text-ratio-200 dark:text-grayBlue-400 cursor-pointer gap-3 mt-3"
    >
      <svg
        fill="none"
        height="60px"
        viewBox="0 0 61 60"
        width="60px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="30.0996"
          cy="30"
          fill="#99B2C6"
          r="30"
        />
        <path
          class="fill-white dark:fill-black-900"
          clip-rule="evenodd"
          d="M26.7216 40.5996C21.1996 40.5996 18.8406 38.2296 18.8406 32.6836V26.5566C18.8406 21.0106 21.2006 18.6416 26.7216 18.6416H32.8215C38.3435 18.6416 40.7025 21.0106 40.7025 26.5566V31.4306C40.7172 31.5167 40.7172 31.6046 40.7025 31.6906V32.6836C40.7025 38.2296 38.3435 40.5996 32.8215 40.5996H26.7216ZM26.7216 39.0676H32.8215C37.5095 39.0676 39.1776 37.3926 39.1776 32.6837V31.9056L37.7955 30.7136C37.2455 30.2743 36.4646 30.2743 35.9146 30.7136L31.6846 34.3596C30.5638 35.2887 28.9404 35.2887 27.8196 34.3596L27.4836 34.0596C26.976 33.6567 26.2675 33.6237 25.7245 33.9777L21.2656 36.9836C22.1995 38.4756 23.9235 39.0726 26.7216 39.0726V39.0676ZM20.3655 26.5566V32.6837C20.3453 33.6467 20.446 34.6086 20.6656 35.5466L24.8806 32.7047C25.9948 31.9654 27.4619 32.0442 28.4905 32.8986L28.8265 33.1986C29.3729 33.6378 30.1512 33.6378 30.6975 33.1986L34.9275 29.5526C36.048 28.6227 37.6719 28.6227 38.7925 29.5526L39.1775 29.8846V26.5596C39.1775 21.8516 37.5095 20.1766 32.8215 20.1766H26.7215C22.0336 20.1786 20.3666 21.8536 20.3666 26.5616L20.3655 26.5566ZM23.9255 25.5356C23.9305 23.9904 25.1863 22.7414 26.7314 22.7447C28.2767 22.748 29.5271 24.0024 29.5255 25.5477C29.5238 27.0929 28.2707 28.3447 26.7255 28.3447C25.982 28.3447 25.269 28.049 24.7438 27.5228C24.2186 26.9966 23.9242 26.2831 23.9255 25.5396V25.5356ZM25.4515 25.5356C25.4543 26.2369 26.0246 26.8035 26.7259 26.8016C27.4272 26.7998 27.9944 26.2303 27.9935 25.5289C27.9926 24.8276 27.4238 24.2596 26.7225 24.2596C26.3842 24.2607 26.0602 24.3962 25.8218 24.6363C25.5834 24.8763 25.4502 25.2013 25.4515 25.5397V25.5356Z"
          fill-rule="evenodd"
        />
      </svg>
      <p
        class="text-center font-normal text-xsm text-ratio-200 dark:text-grayBlue-400"
      >
        Drop your images here, or click to browse 1600 x 1200 (4:3) 
        <br />
        recommended, up to 10MB each.
      </p>
      <div
        class="group flex-col data-[hidden=true]:hidden w-full relative justify-end hidden"
        data-filled="true"
        data-filled-within="true"
        data-has-value="true"
        data-slot="base"
      >
        <div
          class="flex flex-col w-xl h-14"
          data-slot="main-wrapper"
        >
          <div
            class="relative tap-highlight-transparent flex-row shadow-sm gap-3 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 min-h-10 transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background bg-gray-50 flex items-center space-x-2 h-full placeholder:text-ratio-300 dark:bg-gray-700 dark: text-ratio-700 dark:placeholder:text-ratio-700 focus-within:ring-0 py-4 px-6 w-full rounded-lg"
            data-slot="input-wrapper"
            style="cursor: text;"
          >
            <div
              class="w-full items-center h-full box-border flex"
              data-slot="inner-wrapper"
            >
              <input
                accept="image/*"
                aria-label=" "
                class="w-full bg-transparent !outline-none focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text group-data-[has-value=true]:text-default-foreground placeholder:text-ratio-300 dark:placeholder:text-ratio-700 dark: text-ratio-700 flex-grow focus:outline-none focus:ring-0 text-base font-semibold"
                data-slot="input"
                id="react-aria-:r0:"
                multiple=""
                type="file"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="image-previews mt-4 grid grid-cols-2 gap-2"
    />
    <p
      class="font-normal text-xsm text-ratio-200 dark:text-grayBlue-400 mt-3"
    >
      Add up to 10 images to your product. Used to represent your product during checkout, in email, social sharing, and more.
    </p>
  </div>
</div>
`;
